src/ring_theory/finiteness.lean

```diff
@@ -0,0 +1,271 @@
+
+variables {R}
+
+instance prod [hM : finite R M] [hN : finite R N] : finite R (M × N) :=
+begin
+  rw [finite, ← submodule.prod_top],
+  exact submodule.fg_prod hM hN
+end
+
+lemma equiv [hM : finite R M] (e : M ≃ₗ[R] N) : finite R N :=
+of_surjective (e : M →ₗ[R] N) e.surjective
+
+section algebra
+
+lemma trans [algebra A B] [is_scalar_tower R A B] [hRA : finite R A] [hAB : finite A B] :
+  finite R B :=
+let ⟨s, hs⟩ := hRA, ⟨t, ht⟩ := hAB in submodule.fg_def.2
+⟨set.image2 (•) (↑s : set A) (↑t : set B),
+set.finite.image2 _ s.finite_to_set t.finite_to_set,
+by rw [set.image2_smul, submodule.span_smul hs (↑t : set B), ht, submodule.restrict_scalars'_top]⟩
+
+@[priority 100] -- see Note [lower instance priority]
+instance finite_type [hRA : finite R A] : algebra.finite_type R A :=
+subalgebra.fg_of_submodule_fg hRA
+
+end algebra
+
+end finite
+
+end module
+
+namespace algebra
+
+namespace finite_type
+
+lemma self : finite_type R R := ⟨{1}, subsingleton.elim _ _⟩
+
+section
+open_locale classical
+attribute [elab_as_eliminator] mv_polynomial.induction_on'
+protected lemma mv_polynomial (ι : Type*) [fintype ι] : finite_type R (mv_polynomial ι R) :=
+⟨finset.univ.image mv_polynomial.X, begin
+  rw eq_top_iff, refine λ p, mv_polynomial.induction_on' p
+    (λ u x, finsupp.induction u (subalgebra.algebra_map_mem _ x)
+      (λ i n f hif hn ih, _))
+    (λ p q ihp ihq, subalgebra.add_mem _ ihp ihq),
+  rw [add_comm, mv_polynomial.monomial_add_single],
+  exact subalgebra.mul_mem _ ih
+    (subalgebra.pow_mem _ (subset_adjoin $ finset.mem_image_of_mem _ $ finset.mem_univ _) _)
+end⟩
+end
+
+variables {R A B}
+
+lemma of_surjective (hRA : finite_type R A) (f : A →ₐ[R] B) (hf : surjective f) :
+  finite_type R B :=
+begin
+  rw [finite_type] at hRA ⊢,
+  convert subalgebra.fg_map _ f hRA,
+  simpa only [map_top f, @eq_comm _ ⊤, eq_top_iff, alg_hom.mem_range] using hf
+end
+
+lemma equiv (hRA : finite_type R A) (e : A ≃ₐ[R] B) : finite_type R B :=
+hRA.of_surjective e e.surjective
+
+lemma trans [algebra A B] [is_scalar_tower R A B] (hRA : finite_type R A) (hAB : finite_type A B) :
+  finite_type R B :=
+fg_trans' hRA hAB
+
+end finite_type
+
+end algebra
+
+end module_and_algebra
+
+namespace ring_hom
+variables {A B C : Type*} [comm_ring A] [comm_ring B] [comm_ring C]
+
+/-- A ring morphism `A →+* B` is `finite` if `B` is finitely generated as `A`-module. -/
+def finite (f : A →+* B) : Prop :=
+by letI : algebra A B := f.to_algebra; exact module.finite A B
+
+/-- A ring morphism `A →+* B` is of `finite_type` if `B` is finitely generated as `A`-algebra. -/
+def finite_type (f : A →+* B) : Prop := @algebra.finite_type A B _ _ f.to_algebra
+
+namespace finite
+
+variables (A)
+
+lemma id : finite (ring_hom.id A) := module.finite.self A
+
+variables {A}
+
+lemma of_surjective (f : A →+* B) (hf : surjective f) : f.finite :=
+begin
+  letI := f.to_algebra,
+  show module.finite A B,
```

Comment:
I've not checked out the branch, but just looking at the code I'm slightly surprised this `show` line is needed.

Reply:
It's gone now.